### PR TITLE
Use switch expression

### PR DESCRIPTION
**Pull request by drewnoakes**
## Proposed changes

Further experimentation with C# pattern matching.

- Convert switch statements to switch expressions.
- Convert if statements to switch expressions.

## Test methodology 

- Manual review. CI to run unit tests.

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/UserControls/RepoStateVisualiser.cs [hunk 1]
```diff
@@ -53,22 +53,13 @@ internal sealed class RepoStateVisualiser : IRepoStateVisualiser
...
-
-                return workTreeCount != workTreeSubmodulesCount ? Dirty : DirtySubmodules;
-            }
-
-            return workTreeCount == 0 ? Staged : Mixed;
+                0 when workTreeCount == 0 => Clean,
+                0 when notTrackedCount == workTreeCount => UntrackedOnly,
+                0 => workTreeCount != workTreeSubmodulesCount ? Dirty : DirtySubmodules,
+                _ => workTreeCount == 0 ? Staged : Mixed
+            };
```
Comment (ivangrek) [inline GitUI/UserControls/RepoStateVisualiser.cs:62]: Another case for comparison.

```
return (indexCount, workTreeCount) switch
{
    (0, 0) => Clean,
    (0, _) when notTrackedCount == workTreeCount => UntrackedOnly,
    (0, _) when workTreeCount == workTreeSubmodulesCount => DirtySubmodules,
    (0, _) when workTreeCount != workTreeSubmodulesCount => Dirty,
    (_, 0) => Staged,
    (_, _) => Mixed
};
```

Comment (drewnoakes) [reply GitUI/UserControls/RepoStateVisualiser.cs:62]: Tweaked slightly, to:

```c#
            return (indexCount, workTreeCount) switch
            {
                (0, 0) => Clean,
                (0, _) when workTreeCount == notTrackedCount => UntrackedOnly,
                (0, _) when workTreeCount != workTreeSubmodulesCount => Dirty,
                (0, _) => DirtySubmodules,
                (_, 0) => Staged,
                (_, _) => Mixed
            };
```

File: GitUI/CommandsDialogs/FormResolveConflicts.cs [hunk 2]
```diff
@@ -788,25 +788,14 @@ private void HandleSingleSelect()
...
-
-            if (baseFileExists && localFileExists && !remoteFileExists)
-            {
-                conflictDescription.Text = string.Format(_fileModifiedLocallyAndDeletedRemotely.Text, localSide, remoteSide);
-            }
+                (true, true, true) => string.Format(_fileChangeLocallyAndRemotely.Text, localSide, remoteSide),
+                (false, true, true) => string.Format(_fileCreatedLocallyAndRemotely.Text, localSide, remoteSide),
+                (true, false, true) => string.Format(_fileDeletedLocallyAndModifiedRemotely.Text, localSide, remoteSide),
+                (true, true, false) => string.Format(_fileModifiedLocallyAndDeletedRemotely.Text, localSide, remoteSide),
+                _ => null
```
Comment (gerhardol) [inline GitUI/CommandsDialogs/FormResolveConflicts.cs:797]: This overwrites the the default text
```suggestion
                _ => conflictDescription.Text
```

File: GitUI/Theming/Renderers/ComboBoxRenderer.cs [hunk 3]
```diff
@@ -195,20 +195,11 @@ private static void RenderDownArrow(Context ctx, State.DropDown stateid, Rectang
...
 
-            Color arrowColor;
-            switch (stateid)
+            Color arrowColor = stateid switch
             {
-                case State.DropDown.CBXS_DISABLED:
-                    arrowColor = SystemColors.ControlDark;
-                    break;
-
-                // case State.DropDown.CBXS_NORMAL:
```
Comment (gerhardol) [inline GitUI/Theming/Renderers/ComboBoxRenderer.cs:205]: keep the comments?

File: GitUI/Theming/Renderers/HeaderRenderer.cs [hunk 4]
```diff
@@ -75,15 +75,11 @@ public override int GetThemeColor(int ipartid, int istateid, int ipropid, out in
 
         private static Point[] GetArrowPolygon(State.SortArrow stateId, Rectangle prect)
         {
-            switch (stateId)
+            return stateId switch
             {
-                case State.SortArrow.HSAS_SORTEDUP:
-                    return GetUpArrowPolygon(prect);
-
-                // case State.SortArrow.HSAS_SORTEDDOWN:
```
Comment (gerhardol) [inline GitUI/Theming/Renderers/HeaderRenderer.cs:83]: keep comment?

File: GitUI/Theming/Renderers/ScrollBarRenderer.cs [hunk 5]
```diff
@@ -50,41 +50,25 @@ private static Brush GetBackBrush(int stateId, Parts partId)
...
-                        case States.TrackThumb.SCRBS_HOVER:
-                            return SystemBrushes.ControlLight;
-
-                        case States.TrackThumb.SCRBS_HOT:
-                            return SystemBrushes.ControlDark;
-
-                        case States.TrackThumb.SCRBS_PRESSED:
-                            return SystemBrushes.ControlDarkDark;
-
-                        // case States.TrackThumb.SCRBS_DISABLED:
```
Comment (gerhardol) [inline GitUI/Theming/Renderers/ScrollBarRenderer.cs:67]: keep?

File: GitUI/Theming/Renderers/ScrollBarRenderer.cs [hunk 6]
```diff
@@ -50,41 +50,25 @@ private static Brush GetBackBrush(int stateId, Parts partId)
...
                 // case Parts.SBP_SIZEBOX:
                 default:
-                    switch ((States.TrackThumb)stateId)
+                    return (States.TrackThumb)stateId switch
                     {
-                        // case States.TrackThumb.SCRBS_NORMAL:
-                        // case States.TrackThumb.SCRBS_HOVER:
-                        // case States.TrackThumb.SCRBS_HOT:
-                        // case States.TrackThumb.SCRBS_PRESSED:
-                        // case States.TrackThumb.SCRBS_DISABLED:
```
Comment (gerhardol) [inline GitUI/Theming/Renderers/ScrollBarRenderer.cs:84]: keep?

**Review by gerhardol — COMMENTED** [review]
Good improvement
One request for change and some suggestions to keep comments (not all occurrences commented on)

File: GitUI/Theming/Renderers/ScrollBarRenderer.cs (hunk 6 above)
Comment (RussKie) [reply GitUI/Theming/Renderers/ScrollBarRenderer.cs:84]: yes, please

File: GitUI/Theming/Renderers/ComboBoxRenderer.cs (hunk 3 above)
Comment (drewnoakes) [reply GitUI/Theming/Renderers/ComboBoxRenderer.cs:205]: I omitted these intentionally, but can bring them back if there's a strong desire.

There are many other switches on enum values that don't include comments for missing members. For example, in `RenderTextEx`, plenty of the `Parts` enum members are omitted from code and comments.

Comment (drewnoakes) [reply GitUI/Theming/Renderers/ComboBoxRenderer.cs:205]: I omitted these intentionally. 

There are many other switches on enum values that don't include comments for missing members. For example, in `RenderTextEx`, plenty of the `Parts` enum members are omitted from code and comments.

File: GitUI/Theming/Renderers/ScrollBarRenderer.cs (hunk 6 above)
Comment (drewnoakes) [reply GitUI/Theming/Renderers/ScrollBarRenderer.cs:84]: @RussKie as mentioned above I excluded these intentionally. There were quite a few changes here, mostly made via automatic codefix. I personally do not think we need comments for missing values when we have a catch-all clause. What purpose does it serve?

I don't have the apetite to go through and reinstate all these comments manually, especially given the StyleCop formatting constraints around comments. If you prefer comments, let's just close this out to keep them.

Comment (drewnoakes) [reply GitUI/Theming/Renderers/ScrollBarRenderer.cs:84]: This particular switch expression is a n-op and should just be inlined anyway.

**Review by RussKie — APPROVED** [review]

**Pull request merged**